### PR TITLE
Closes #2241: Use Keccak-256 from golang.org/x/crypto/sha3 and mention explicitly

### LINKED ISSUE #2241: Old & copied SHA-3 code and misleading reference to SHA-3 instead of Keccak-256
This blog explains the issue: https://medium.com/@ConsenSys/are-you-really-using-sha-3-or-old-code-c5df31ad2b0. Ethereum uses **Keccak-256**, not **SHA-3**, which was standardized afterwards. 

I could myself prepare a PR with this improvements:
- Add to Godeps `golang.org/x/crypto/sha3` with a small patch to expose Keccak-256: https://github.com/jimenezrick/crypto/commit/8802e1c340c41c8d290e6315609cfbcbdf1b6229 (this thin fork could be hosted in the Ethereum Github obviously)
- Using upstream `crypto/sha3` gives some *extra performance*, tested running go test benchmark.
- Remove old copied **SHA-3 code**.
- Rename any code referencing **SHA-3** instead of **Keccak-256**

### PR DESCRIPTION
**Pull request by jimenezrick**
First of all, apologies for this long PR.

These are the improvements I propose:
- Use latest version of `golang.org/x/crypto/sha3` with a small patch to expose **Keccak-256** and update the old copied code. The upstream version of sha3 is a bit faster.
- Rename all the misleading occurrences of **SHA-3** to **Keccak-256**
- Remove some bits of dead/old code I saw while doing this

~~Before merging there are two things that need to be taken care:~~
- ~~The `golang.org/x/crypto` fork (with just this modification: https://github.com/jimenezrick/crypto/commit/8802e1c340c41c8d290e6315609cfbcbdf1b6229) should be hosted somewhere, it could be the Ethereum Github: something like github.com/ethereum/go-crypto - *that new import path could be used instead*~~
- ~~The Go code in https://github.com/ethereum/ethash needs to be updated remove references to: `crypto.Sha3{,Hash}()`~~

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Sun Feb 21 21:28:25 UTC 2016_

**Comment by jimenezrick** [comment]
The builds will fail until the necessary update to `ethash` I mentioned before is applied.

File: common/bytes.go [hunk 1]
```diff
@@ -14,7 +14,6 @@
 // You should have received a copy of the GNU Lesser General Public License
 // along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
 
-// Package common contains various helper functions.
```
Comment (fjl) [inline common/bytes.go]: This "unhelpful comment" is the Godoc synopsis of package common and should not be deleted.

Comment (jimenezrick) [inline common/bytes.go]: No problem, line preserved.

**Comment by fjl** [comment]
While I certainly appreciate the effort you took to clean up the code, there are a few issues with the PR that I would like to see resolved. The main issue is that we generally avoid vendoring a patched version of a dependency with godep, because it will break as soon as someone else updates the dependencies from upstream.

So far, we have treated crypto dependencies specially by vendoring them directly into our tree, for various reasons. I think it would be better to import the upstream changes into `crypto/sha3` and then patch the version in our tree instead.

**Comment by jimenezrick** [comment]
Yeah, that sounds right. It's a bit of a mess to vendor in Go modified libraries. I can adapt the PR easily to do that instead.

**Comment by fjl** [comment]
For crypto/sha3, the last update from upstream happened in June 2015. Since then the only change has been to include the full test of the license in all files. Please make sure that they stay in place.

**Comment by fjl** [comment]
For the ethash issue, I think it might be better to preserve `crypto.Sha3{,Hash}` as an alias for now.
We can do a second pass and remove the alias later.

**Comment by fjl** [comment]
Ahh, right, one more thing: the commit convention in go-ethereum is to prefix commit messages with the package they modify, e.g. "bytes: remove dead code". If you do a change affecting a lot of packages, you can use prefix "all:".

**Comment by jimenezrick** [comment]
Roger that, cheers.

**Comment by fjl** [comment]
Thank you for contributing.

**Comment by jimenezrick** [comment]
- I added an alias so that it doesn't break `ethash`
- Upstream `sha3` code copied verbatim, including the LICENSE and PATENTS files and small modification to expose `Keccak256`
- Reorder of all the commits 

@fjl thanks for the suggestions, it's way more clear now ;)

**Comment by fjl** [comment]
:+1:

**Comment by obscuren** [comment]
Thanks for the PR @jimenezrick, appreciated. :+1:

**Pull request merged**